[SPARK-40640][CORE] SparkHadoopUtil to set origin of hadoop/hive config options

steveloughran opened the pull request:
### What changes were proposed in this pull request?

The options passed from spark conf, hive-site.xml, AWS env vars now all record this in their source attribute of the entries.

The Configuration Writable methods do not propagate this, so it is not as useful cluster-wide than it could be. It does help with some of the basic troubleshooting.


### Why are the changes needed?
 
Helps when troubleshooting where options make their way down. These can be examined
and logged later.

For example, my cloudstore diagnosticss JAR can do this in its storediag command
and in an s3a AWS credential provider. I may add some of that logging
at debug to the ASF hadoop implementations.

https://github.com/steveloughran/cloudstore

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

Not *really*. It's a very low level diagnostics feature in the Hadoop configuration classes.

### How was this patch tested?

New tests added; existing tests enhanced.

steveloughran committed:
SPARK-40640. SparkHadoopUtil to set origin of hadoop/hive config options

The options passed from spark conf, hive-site.xml, AWS env vars
now all record this in their source attribute of the entries.

The Configuration Writable methods do not propagate this,
so it is not as useful cluster-wide than it could be.
It does help with some of the basic troubleshooting.

Change-Id: Id3c7cbf1f1d17ef1bf3cc7d8144c84260861a4e5

attilapiros commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -437,27 +437,31 @@ private[spark] object SparkHadoopUtil extends Logging {
-        hadoopConf.set("fs.s3a.secret.key", accessKey)
-
-        val sessionToken = System.getenv("AWS_SESSION_TOKEN")
+        val source = "Spark set from "
```
```suggestion
        val source = "Set by Spark from "
```

attilapiros commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -93,4 +123,42 @@ class SparkHadoopUtilSuite extends SparkFunSuite {
+    // get the possibly null source list
+    val sources = hadoopConf.getPropertySources(key)
+    assert(sources != null && sources.length > 0,
+      s"Expected source for $key with value $v")
```
```suggestion
      s"Sources are missing for '$key' with value '$v'")
```

steveloughran commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -474,21 +478,22 @@ private[spark] object SparkHadoopUtil extends Logging {
+    val launcher = "spark launch"
     if (conf.getOption("spark.hadoop.mapreduce.fileoutputcommitter.algorithm.version").isEmpty) {
-      hadoopConf.set("mapreduce.fileoutputcommitter.algorithm.version", "1")
+      hadoopConf.set("mapreduce.fileoutputcommitter.algorithm.version", "1", launcher)
```
ok. i don't personally have any opinions on what is best, just knowing the origin is useful

steveloughran committed:
Update core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

Co-authored-by: Attila Zsolt Piros <2017933+attilapiros@users.noreply.github.com>

steveloughran committed:
Update core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

Co-authored-by: Attila Zsolt Piros <2017933+attilapiros@users.noreply.github.com>

steveloughran committed:
Update core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

Co-authored-by: Attila Zsolt Piros <2017933+attilapiros@users.noreply.github.com>

steveloughran committed:
Update core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

Co-authored-by: Attila Zsolt Piros <2017933+attilapiros@users.noreply.github.com>

steveloughran committed:
Update core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala

Co-authored-by: Attila Zsolt Piros <2017933+attilapiros@users.noreply.github.com>

steveloughran committed:
SPARK-40640. SparkHadoopUtil to set origin of hadoop/hive config options

* update tests to match changed strings
* HiveClientImpl.newHiveConf() does some of this too.
  It doesn't pass on the origin from any hadoop conf passed in because
  that conf comes in as JIterable[JMap.Entry[String, String]];
  reflection could be used *or* the params changed, which
  introduces risk to what is just a minor debug change.
* SparkHadoopUtil.extractOrigin() is there for anyone
  who ever wants to do that in future.

Change-Id: Ia961c6b084e4cdba7e55d5687b32d9ea14c73b84

steveloughran committed:
SPARK-40640. SparkHadoopUtil to set origin of hadoop/hive config options

remove the unused ref to SparkHadoopUtil in HiveClientImpl

Change-Id: Id604178023762f8712dc727073232f6dcf5ca7cd

attilapiros commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -500,14 +523,35 @@ private[spark] object SparkHadoopUtil extends Logging {
+   * @param defVal default value
+   * @return the origin of the current entry in the configuration, or the default.
+   */
+  def extractOrigin(hadoopConf: Configuration, key: String, defVal: String): String = {
```
I suggest to use the name `propertySource` (or `propertySources` if all is needed see below) to decrease the need of mapping of names.

attilapiros replied:
So do we know something about the orders? 
1) Which one wins if one key is set from two different sources? The last one?
2) Does the last on is the first one in the `getPropertySources()` result?

attilapiros replied:
If we have the answer we should fix the documentation (in the Hadoop code).

steveloughran commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -105,6 +105,23 @@ private[spark] class SparkHadoopUtil extends Logging {
+   * @param defVal     default value
+   * @return the origin of the current entry in the configuration, or the default.
+   */
+  def extractOrigin(hadoopConf: Configuration, key: String,
```
ok; i was just going with the rest

steveloughran commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -500,14 +523,35 @@ private[spark] object SparkHadoopUtil extends Logging {
+  def extractOrigin(hadoopConf: Configuration, key: String, defVal: String): String = {
+    val sources = hadoopConf.getPropertySources(key)
+    if (sources != null && sources.length > 0) {
+      sources(0)
```
I'll do the mkstring(); it can only help debugging to know that entire chain. This is what cloudstore does, incidentally. I also plan to improve the s3a credential providers to log at debug their sources as passed in, maybe even some simple crc2 checksum (so you can use crc32 on the command like to verify your secrets are in, but not enough that if someone has the log they can determine a valid set of credentials)

steveloughran committed:
SPARK-40640. SparkHadoopUtil to set origin of hadoop/hive config options

1. Review comments
2. AWS env var propagation records hostname on which these were added.
   This helps identify issues where configurations are being created in
   the cluster, rather than on the deploying host -the one with the secrets.
3. extracted the env propagation method into its own private[deploy] method
   so the unit tests can validate this works.
4, reviewed comments on other setters and updated them to explicitly
   state which hadoop versions have the issues which need fixing.

Change-Id: Ie58856ce621e288c4a8bb06083e43f450f4ec26d

steveloughran commented:
one more idea. To help debug the origin of config files, how about we always patch the config with (if unset) some property like spark.diagnostics.config.origin to that hostname?
that way, if you ever get a config in your debugger, you can print that value and see where it originated?

attilapiros commented:
> one more idea. To help debug the origin of config files, how about we always patch the config with (if unset) some property like spark.diagnostics.config.origin to that hostname? that way, if you ever get a config in your debugger, you can print that value and see where it originated?

Ok but let's do that in separate PR.

steveloughran commented:
that last change seems to break a yarn test
```
- yarn-cluster should respect conf overrides in SparkHadoopUtil (SPARK-16414, SPARK-23630) *** FAILED *** (3 minutes, 0 seconds)
[info]   The code passed to eventually never returned normally. Attempted 190 times over 3.001129006083333 minutes. Last failure message: handle.getState().isFinal() was false. (BaseYarnClusterSuite.scala:209)
[info]   org.scalatest.exceptions.TestFailedDueToTimeoutException:
[info]   at org.scalatest.enablers.Retrying$$anon$4.tryTryAgain$2(Retrying.scala:219)
[info]   at org.scalatest.enablers.Retrying$$anon$4.retry(Retrying.scala:226)
[info]   at org.scalatest.concurrent.Eventually.eventually(Eventually.scala:313)
[info]   at org.scalatest.concurrent.Eventually.eventually$(Eventually.scala:312)
[info]   at org.scalatest.concurrent.Eventually$.eventually(Eventually.scala:457)
[info]   at org.apache.spark.deploy.yarn.BaseYarnClusterSuite.runSpark(BaseYarnClusterSuite.scala:209)
[info]   at org.apache.spark.deploy.yarn.YarnClusterSuite.$anonfun$new$6(YarnClusterSuite.scala:148)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.apache.spark.deploy.yarn.BaseYarnClusterSuite.$anonfun$test$1(BaseYarnClusterSuite.scala:77)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
```

attilapiros commented:
Interesting locally (on its own) it was passing:
```
➜  spark-pr1 git:(pr/38084) ✗ git log --oneline -n 1
7909817173 (HEAD -> pr/38084) SPARK-40640. SparkHadoopUtil to set origin of hadoop/hive config options

➜  spark-pr1 git:(pr/38084) ✗ ./build/sbt -Pyarn 'project yarn; testOnly *.YarnClusterSuite -- -z "yarn-cluster should respect conf overrides in SparkHadoopUtil"'
....
[info] compiling 4 Scala sources to /Users/attilazsoltpiros/git/attilapiros/spark-pr1/resource-managers/yarn/target/scala-2.12/test-classes ...
[info] YarnClusterSuite:
[info] - yarn-cluster should respect conf overrides in SparkHadoopUtil (SPARK-16414, SPARK-23630) (13 seconds, 231 milliseconds)
[info] Run completed in 34 seconds, 440 milliseconds.
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 59 s, completed Oct 6, 2022 2:15:07 PM
sbt:spark-yarn>
[info] shutting down sbt server
```

attilapiros commented:
Those are flaky tests:
https://issues.apache.org/jira/browse/SPARK-32297

Please re-trigger the check!

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -18,9 +18,11 @@
 import java.security.PrivilegedExceptionAction
 import java.text.DateFormat
 import java.util.{Arrays, Date, Locale}
+import java.util
```
```scala
-import java.util.{Arrays, Date, Locale}
-import java.util
+import java.util.{Arrays, Date, Locale, Map => JMap}
```

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -438,26 +440,48 @@ private[spark] object SparkHadoopUtil extends Logging {
-          hadoopConf.set("fs.s3a.session.token", sessionToken)
-        }
-      }
+      val env: util.Map[String, String] = System.getenv
```
```scala
-      val env: util.Map[String, String] = System.getenv
+      val env: JMap[String, String] = System.getenv
```

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -438,26 +440,48 @@ private[spark] object SparkHadoopUtil extends Logging {
+  // Exposed for testing
+  private[deploy] def appendS3CredentialsFromEnvironment(
+    hadoopConf: Configuration,
+    env: util.Map[String, String]): Unit = {
```
```scala
-    hadoopConf: Configuration,
-    env: util.Map[String, String]): Unit = {
+      hadoopConf: Configuration,
+      env: JMap[String, String]): Unit = {
```

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -438,26 +440,48 @@ private[spark] object SparkHadoopUtil extends Logging {
-        }
-      }
+      val env: util.Map[String, String] = System.getenv
+      appendS3CredentialsFromEnvironment(hadoopConf, env)
```
If `env` is used by only `appendS3CredentialsFromEnvironment`, shall we invoke `System.genenv` inside that method instead of here.

steveloughran replied:
i want it isolated for testing; we can't set env vars in a unit test, but we can instead build up a map and verify propagation there.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -474,40 +498,79 @@ private[spark] object SparkHadoopUtil extends Logging {
+      hadoopConf.set(key.substring("spark.hadoop.".length), value,
+        "Set by Spark from keys starting with 'spark.hadoop'")
     }
+    val setBySpark = "Set by Spark to default values"
```
Could you define these constants in `object SparkHadoopUtil`, please?
- `"Set by Spark from hive-site.xml"`
- `"Set by Spark to default values"`

steveloughran replied:
happy to

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -17,12 +17,19 @@
 
+  private val setToDefaultValues = "Set by Spark to default values"
+  private val hadoopPropagation = "Set by Spark from keys starting with 'spark.hadoop'"
+  private val hivePropagation = "Set by Spark from keys starting with 'spark.hive'"
```
We can remove these by https://github.com/apache/spark/pull/38084/files#r989713735.

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -80,6 +89,43 @@ class SparkHadoopUtilSuite extends SparkFunSuite {
+  /**
+   * The explicit buffer size propagation records this.
+   */
+  test("buffer size propagation") {
```
```scala
- test("buffer size propagation") {
+ test("SPARK-40640: buffer size propagation") {
```

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -80,6 +89,43 @@ class SparkHadoopUtilSuite extends SparkFunSuite {
+    assertConfigMatches(hadoopConf, "io.file.buffer.size", "123", BUFFER_SIZE.key)
+  }
+
+  test("aws credentials from environment variables") {
```
```scala
- test("aws credentials from environment variables") {
+ test("SPARK-40640: aws credentials from environment variables") {
```

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -93,4 +139,41 @@ class SparkHadoopUtilSuite extends SparkFunSuite {
+    hadoopConf: Configuration,
+    key: String,
+    expected: String,
+    expectedSource: String): Unit = {
```
Indentation: Two more spaces.
```scala
  private def assertConfigMatches(
       hadoopConf: Configuration,
       key: String,
       expected: String,
       expectedSource: String): Unit = {
```

steveloughran replied:
will need to tweak my scala ide rules, i see

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -93,4 +139,41 @@ class SparkHadoopUtilSuite extends SparkFunSuite {
+  private def assertConfigSourceContains(
+    hadoopConf: Configuration,
+    key: String,
+    expectedSource: String): Unit = {
```
ditto. Indentation. Two more spaces for parameters.

dongjoon-hyun commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -1299,7 +1299,8 @@ private[hive] object HiveClientImpl extends Logging {
     val confMap = (hadoopConf.iterator().asScala.map(kv => kv.getKey -> kv.getValue) ++
       sparkConf.getAll.toMap ++ extraConfig).toMap
-    confMap.foreach { case (k, v) => hiveConf.set(k, v) }
+    val fromSpark = "Set by Spark"
```
Although this is tricky, let's collect this into the same place (https://github.com/apache/spark/pull/38084/files#r989713735).

dongjoon-hyun replied:
Gentle ping about this leftover.

dongjoon-hyun replied:
Since this is non-test code string, we should have in the same place with the others.

steveloughran replied:
yes, what is the solution there? Even when exported with private[spark] it wasn't picked up

dongjoon-hyun replied:
> yes, what is the solution there? Even when exported with private[spark] it wasn't picked up

Okay, let me make a PR to you.

dongjoon-hyun reviewed (commented):
I finished the first round review, @steveloughran . In general, the PR looks helpful.

steveloughran committed:
SPARK-40640. review feedback

* pull out source strings and env vars into constants; use
* add a new test case

Change-Id: I88f525745c046cc7040eab5501449447a0d5924c

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -474,40 +553,79 @@ private[spark] object SparkHadoopUtil extends Logging {
     for ((key, value) <- conf.getAll if key.startsWith("spark.hive.")) {
-      hadoopConf.set(key.substring("spark.".length), value)
+      hadoopConf.set(key.substring("spark.".length), value,
+        SOURCE_SPARK_HIVE)
```
Can we make it into a single line?
```scala
hadoopConf.set(key.substring("spark.".length), value, SOURCE_SPARK_HIVE)
```

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -437,27 +490,53 @@ private[spark] object SparkHadoopUtil extends Logging {
-          hadoopConf.set("fs.s3a.session.token", sessionToken)
-        }
-      }
+      appendS3CredentialsFromEnvironment(hadoopConf, System.getenv)
```
I'm still worrying about the chance of the regression of getting and passing a large map of strings here. There were some reports like this.
- https://github.com/apache/spark/pull/31244/files

dongjoon-hyun replied:
If you really want to hand over a map as a dependency injection, can you build a small one by reusing the existing logic only like `System.getenv("AWS_ACCESS_KEY_ID")`?

steveloughran replied:
ok

steveloughran replied:
having it pass down separate arguments; no map at all

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -474,40 +553,79 @@ private[spark] object SparkHadoopUtil extends Logging {
+
+  /**
+   * Return a hostname without throwing an exception if the system
+   * does not know its own name.
```
I'm wondering when does this happen? Apache Spark has the following codes.

```
core/src/main/scala/org/apache/spark/util/Utils.scala:      val address = InetAddress.getLocalHost
core/src/main/scala/org/apache/spark/util/Utils.scala:            logWarning("Your hostname, " + InetAddress.getLocalHost.getHostName + " resolves to" +
core/src/main/scala/org/apache/spark/util/Utils.scala:        logWarning("Your hostname, " + InetAddress.getLocalHost.getHostName + " resolves to" +
```

steveloughran replied:
that is an interesting q. tracing the history from hadoop NetUtil and back, came in with https://issues.apache.org/jira/browse/HADOOP-1028 *and no explicit "we have seen this". I have seen machines with the wrong name, wrong IP, and even localhost not mapped to a 127. address. but not this. will cut and assume nicholas just added it to save on exception propagation in java

steveloughran replied:
inlining

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -32,9 +36,12 @@ class SparkHadoopUtilSuite extends SparkFunSuite {
-    assertConfigValue(hadoopConf, "fs.s3a.downgrade.syncable.exceptions", "true")
-    assertConfigValue(hadoopConf, "fs.s3a.endpoint", "s3.amazonaws.com")
+    assertConfigMatches(hadoopConf, "orc.filterPushdown", "true",
+      SOURCE_SPARK_HADOOP)
```
Do we need to split the line? Otherwise, please make it as a single line.

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -80,17 +87,100 @@ class SparkHadoopUtilSuite extends SparkFunSuite {
+  /**
+   * spark.hive.* is passed to the hadoop config as hive.*.
+   */
+  test("spark.hive propagation") {
```
Please add a test case prefix, `SPARK-40640: `.

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/SparkHadoopUtilSuite.scala:
```diff
@@ -80,17 +87,100 @@ class SparkHadoopUtilSuite extends SparkFunSuite {
+    assertConfigMatches(hadoopConf, "fs.s3a.session.token", "session-token", source)
+  }
+
+  test("SPARK-19739: S3 session token propagation requires access and secret keys") {
```
I understand what you mean, but you should not add irrelevant test cover in this PR. You don't use `assertConfigMatches` or `assertConfigSourceContains` here. So, it looks like completely independent test case. Please proceed this as a separate test case PR.

steveloughran replied:
its more now test is doing env var probing, it's verifying that it doesn't pick up the session ids. but if you don't want it in, i will cut

dongjoon-hyun replied:
Got it. New one looks good to me too.

steveloughran committed:
SPARK-40640. review feedback -almost all comments.

* move from env var map to explicit args of values; test gets leaner

Change-Id: I74f76427321296583235c9600b10a70cbd42777a

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -474,40 +559,61 @@ private[spark] object SparkHadoopUtil extends Logging {
+  def propertySources(hadoopConf: Configuration, key: String): String = {
+    val sources = hadoopConf.getPropertySources(key)
+    if (sources != null && sources.nonEmpty) {
+      sources.mkString("," )
```
nit.
```scala
- sources.mkString("," )
+ sources.mkString(",")
```

dongjoon-hyun committed:
Minor change

dongjoon-hyun reviewed (changes requested):
I made a PR to your branch, @steveloughran .
- https://github.com/steveloughran/spark/pull/1

steveloughran committed:
Merge pull request #1 from dongjoon-hyun/SPARK-PR-38084

Minor changes

steveloughran commented:
thanks, merged in. when i tried that xref of the constant to hive class things wouldn't resolve properly; clearly some local build issue.

steveloughran commented:
test failure seems nowhere near this code
```
[info] *** 1 TEST FAILED ***
[error] Failed: Total 3425, Failed 1, Errors 0, Passed 3424, Ignored 9, Canceled 2
[error] Failed tests:
[error] 	org.apache.spark.scheduler.TaskSetManagerSuite
```

dongjoon-hyun commented:
Ya, I agree with you.
```
[info] - SPARK-32170: test SPECULATION_EFFICIENCY_TASK_DURATION_FACTOR for speculating tasks *** FAILED *** (106 milliseconds)
[info]   manager.checkSpeculatableTasks(15000L) was true (TaskSetManagerSuite.scala:2487)
```

dongjoon-hyun reviewed (approved):
+1, LGTM. Merged to master for Apache Spark 3.4.0.

steveloughran commented:
thx. still not sufficient, but it may help.
